sub/sd_sbr.c

```diff
@@ -0,0 +1,229 @@
+};
+
+static void enable_output(struct sd *sd, bool enable)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    if (enable == !!ctx->sbr_renderer)
+        return;
+    if (ctx->sbr_renderer) {
+        sbr_renderer_destroy(ctx->sbr_renderer);
+        ctx->sbr_renderer = NULL;
+    } else {
+        ctx->sbr_renderer = sbr_renderer_create(ctx->sbr_library);
+        mp_require(ctx->sbr_renderer);
+    }
+}
+
+static inline int mp_level_from_sbr_log_level(sbr_log_level level)
+{
+    switch (level) {
+        case SBR_LOG_LEVEL_TRACE:
+            return MSGL_TRACE;
+        case SBR_LOG_LEVEL_DEBUG:
+            return MSGL_DEBUG;
+        case SBR_LOG_LEVEL_INFO: // fallthrough
+        case SBR_LOG_LEVEL_WARN:
+            return MSGL_V;
+        case SBR_LOG_LEVEL_ERROR: // fallthrough
+        default:
+            return MSGL_WARN;
+    }
+}
+
+static void mp_msg_sbr_log_callback(sbr_log_level level,
+                                    const char *source, size_t source_len,
+                                    const char *message, size_t message_len,
+                                    void *user_data)
+{
+    struct sd *sd = user_data;
+    int mp_level = mp_level_from_sbr_log_level(level);
+
+    if (mp_msg_test(sd->log, mp_level)) {
+        if (source_len > 0)
+            mp_msg(sd->log, mp_level, "[%.*s] ", (int)source_len, source);
+        mp_msg(sd->log, mp_level, "%.*s\n", (int)message_len, message);
+    }
+}
+
+static int init(struct sd *sd)
+{
+    if (!sd->codec->codec)
+        return -1;
+    if (strcmp(sd->codec->codec, "subrandr/srv3") && strcmp(sd->codec->codec, "subrandr/vtt"))
+        return -1;
+
+    struct sd_sbr_priv *ctx = talloc_zero(sd, struct sd_sbr_priv);
+    sd->priv = ctx;
+
+    ctx->sbr_library = sbr_library_init();
+    mp_require(ctx->sbr_library);
+    sbr_library_set_log_callback(ctx->sbr_library, mp_msg_sbr_log_callback, sd);
+
+    ctx->bitmaps = talloc_zero(ctx, struct sub_bitmaps);
+    ctx->bitmaps->format = SUBBITMAP_BGRA;
+    ctx->bitmaps->num_parts = 1;
+    ctx->bitmaps->parts = talloc_zero(ctx->bitmaps, struct sub_bitmap);
+
+    enable_output(sd, true);
+
+    return 0;
+}
+
+static void decode(struct sd *sd, struct demux_packet *packet)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+
+    if (ctx->sbr_subtitles)
+        sbr_subtitles_destroy(ctx->sbr_subtitles);
+    if (ctx->sbr_renderer)
+        sbr_renderer_set_subtitles(ctx->sbr_renderer, NULL);
+
+    sbr_subtitle_format fmt = SBR_SUBTITLE_FORMAT_UNKNOWN;
+    if (!strcmp(sd->codec->codec, "subrandr/srv3") )
+        fmt = SBR_SUBTITLE_FORMAT_SRV3;
+    else if (!strcmp(sd->codec->codec, "subrandr/vtt"))
+        fmt = SBR_SUBTITLE_FORMAT_WEBVTT;
+
+    ctx->sbr_subtitles = sbr_load_text(
+        ctx->sbr_library,
+        packet->buffer,
+        packet->len,
+        fmt,
+        NULL
+    );
+    packet->sub_duration = packet->duration;
+
+    if (!ctx->sbr_subtitles) {
+        const char *error = sbr_get_last_error_string();
```

Comment:
btw I hope this is thread-safe?

Reply:
yes the last error is kept in a thread local and this function returns the *thread's* last error.